src/Features/Core/Portable/Completion/Providers/TypeImportCompletionService.cs

```diff
@@ -0,0 +1,296 @@
+            Solution solution,
+            Compilation compilation,
+            PortableExecutableReference peReference,
+            Action<TypeImportCompletionItemInfo> handleAccessibleItem,
+            CancellationToken cancellationToken);
+    }
+
+    internal readonly struct TypeImportCompletionItemInfo
+    {
+        public TypeImportCompletionItemInfo(CompletionItem item, bool isPublic)
+        {
+            Item = item;
+            IsPublic = isPublic;
+        }
+
+        public CompletionItem Item { get; }
+
+        public bool IsPublic { get; }
+    }
+
+    [ExportWorkspaceServiceFactory(typeof(ITypeImportCompletionService), ServiceLayer.Editor), Shared]
+    internal sealed class TypeImportCompletionService : IWorkspaceServiceFactory
+    {
+        private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache
+            = new ConcurrentDictionary<string, ReferenceCacheEntry>();
+
+        private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache
+            = new ConcurrentDictionary<ProjectId, ReferenceCacheEntry>();
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+        {
+            var cacheService = workspaceServices.GetService<IWorkspaceCacheService>();
+            if (cacheService != null)
+            {
+                cacheService.CacheFlushRequested += OnCacheFlushRequested;
+            }
+
+            return new Service(_peItemsCache, _projectItemsCache);
+        }
+
+        private void OnCacheFlushRequested(object sender, EventArgs e)
+        {
+            _peItemsCache.Clear();
+            _projectItemsCache.Clear();
+        }
+
+        private class Service : ITypeImportCompletionService
+        {
+            // PE references are keyed on assembly path.
+            private readonly ConcurrentDictionary<string, ReferenceCacheEntry> _peItemsCache;
+            private readonly ConcurrentDictionary<ProjectId, ReferenceCacheEntry> _projectItemsCache;
+
+            public Service(ConcurrentDictionary<string, ReferenceCacheEntry> peReferenceCache, ConcurrentDictionary<ProjectId, ReferenceCacheEntry> projectReferenceCache)
+            {
+                _peItemsCache = peReferenceCache;
+                _projectItemsCache = projectReferenceCache;
+            }
+
+            public async Task GetTopLevelTypesFromProjectAsync(
+                Project project,
+                Action<TypeImportCompletionItemInfo> handleItem,
+                CancellationToken cancellationToken)
+            {
+                if (!project.SupportsCompilation)
+                {
+                    throw new ArgumentException(nameof(project));
+                }
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
```

Comment:
needs comment about this being the right checksum.  should be simple, since we're only creating items for the top level source types, then we only care if the source symbold checksum changes.

Reply:
Finally added after you commented on this like 5 times :)

Reply:
:D